[FLINK-25105][checkpoint] Enables final checkpoint by default (#18068)
apache/flink: docs/content/docs/dev/datastream/fault-tolerance/checkpointing.md

  - *unaligned checkpoints*: You can enable [unaligned checkpoints]({{< ref "docs/ops/state/checkpointing_under_backpressure" >}}#unaligned-checkpoints) to greatly reduce checkpointing times under backpressure. This only works for exactly-once checkpoints and with one concurrent checkpoint.

>   - *checkpoints with finished tasks*: By default Flink could continue performing checkpoints even if parts of the DAG have finished processing all of their records. Please refer to [important considerations](#checkpointing-with-parts-of-the-graph-finished) for details.

Reviewer: ```suggestion
  - *checkpoints with finished tasks*: By default Flink will continue performing checkpoints even if parts of the DAG have finished processing all of their records. Please refer to [important considerations](#checkpointing-with-parts-of-the-graph-finished) for details.
```

Please note that records in flight in the loop edges (and the state changes associated with them) will be lost during failure.

> ## Checkpointing with parts of the graph finished

Reviewer: Do we have documented somewhere that finished operators are waiting for the final checkpoint to complete before closing? If not, can you add such section that? It should also explicitly state that it can prolong completion of the job, until that last final checkpoint is triggered, so longer checkpointing intervals can block job from completing.

Starting from Flink 1.14 it is possible to continue performing checkpoints even if parts of the job
graph have finished processing all data, which might happen if it contains bounded sources. This feature

> is enabled by default since 1.15, and it could be configured by

Reviewer: ```suggestion
is enabled by default since 1.15, and it could be disabled via a feature flag:
```

```java
Configuration config = new Configuration();

> config.set(ExecutionCheckpointingOptions.ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH, true);

Reviewer: ```suggestion
config.set(ExecutionCheckpointingOptions.ENABLE_CHECKPOINTS_AFTER_TASKS_FINISH, false);
```

Restoring from a checkpoint where only a subset of tasks finished is equivalent to restoring such a
task with the number of new subtasks equal to the number of finished tasks.

### Waiting for the final checkpoint before task exit

> To ensure all the records could be committed for operators using the two-phase commit, the tasks would wait for the final checkpoint completed successfully after all the operators finished. It needs to be noted that this behavior would prolong the execution time of tasks. If the checkpoint interval is long, the execution time would also be prolonged largely. For the worst case, if the checkpoint interval is set to `Long.MAX_VALUE`, the tasks would in fact be blocked forever since the final checkpoint would never happen.

Reviewer: ```
To ensure all the records will be committed for operators using the two-phase commit, the tasks have to wait for the final checkpoint to complete successfully after all operators has finished processing new records. It is worth noting that this behavior prolongs the execution time of tasks. If the either checkpoint interval is long or end to end checkpoint duration is long, the execution time will also be prolonged significantly. In the worst case, if the checkpoint interval is set to `Long.MAX_VALUE`, the tasks would in fact be blocked forever since the final checkpoint will never happen.
```

+ can you split it into multiple lines? Docs that have paragraphs in a single line are causing quite a huge mess when reviewing the diffs or resolving merge conflicts.

Author: I split the lines now~
